### ISSUE TITLE
Private channel not working

### ISSUE THREAD

**Issue by xparthxvorax** [tlaverdure/laravel-echo-server#158]
Hello,

I'm getting this JS error on the console:
app.js:167 Uncaught ReferenceError: receiverId is not defined

**Here is my complete code:**

_PrivateChatController:_

`event(new PrivateMessageEvent($chat, $receiverId));`

_PrivateMessageEvent:_

```
namespace App\Events;

use Illuminate\Broadcasting\Channel;
use Illuminate\Queue\SerializesModels;
use Illuminate\Broadcasting\PrivateChannel;
use Illuminate\Broadcasting\PresenceChannel;
use Illuminate\Foundation\Events\Dispatchable;
use Illuminate\Broadcasting\InteractsWithSockets;
use Illuminate\Contracts\Broadcasting\ShouldBroadcast;
use App\User;
use App\PrivateChat;

class PrivateMessageEvent implements ShouldBroadcast
{
    use Dispatchable, InteractsWithSockets, SerializesModels;

    public $privateChat, $receiverId;
    /**
     * Create a new event instance.
     *
     * @return void
     */
    public function __construct(PrivateChat $privateChat, $receiverId)
    {
        $this->privateChat = $privateChat;
        $this->receiverId = $receiverId;
    }

    /**
     * Get the channels the event should broadcast on.
     *
     * @return Channel|array
     */
    public function broadcastOn()
    {
        return new PrivateChannel('private-chat.' . $this->receiverId);
    }
}
```
_Bootstrap.js_

```
import Echo from "laravel-echo"

window.Echo = new Echo({
    broadcaster: 'socket.io',
    host: window.location.hostname + ':6001'
});

window.Echo.private(`private-chat.${receiverId}`)
    .listen('PrivateMessageEvent', (e) => {
		console.log(e);
});
```

_channels.php_

```
Broadcast::channel('private-chat.{receiverId}', function ($user, $receiverId) {
    return true; // this is just for debugging to allow anyone to listen on this channel
    //return $user->id === $receiverId;
});
```

_laravel-echo-server.json_

```
{
	"authHost": "http://localhost",
	"authEndpoint": "/broadcasting/auth",
	"clients": [],
	"database": "redis",
	"databaseConfig": {
		"redis": {},
		"sqlite": {
			"databasePath": "/database/laravel-echo-server.sqlite"
		}
	},
	"devMode": true,
	"host": null,
	"port": "6001",
	"protocol": "http",
	"socketio": {},
	"sslCertPath": "",
	"sslKeyPath": ""
}
```

In background _queue:work_ and _laravel-echo-server_ are already running

Upon firing that event, I'm getting this message on the laravel-echo-server console:

```
Channel: private-private-chat.
Event: App\Events\PrivateMessageEvent
CHANNEL private-private-chat.
```

**Notes:**

- I'm successfully able to listen to the public channel. The only issue with the private channel.
- Using latest Laravel version i.e 5.4
- I have done all the things as per the official docs:
  https://laravel.com/docs/master/broadcasting
  https://github.com/tlaverdure/laravel-echo-server

I have spent more than 10 hours and tried everything I could, but no luck.

Thanks

**Comment by kkomelin** [comment]
Hi @xparthx,

> app.js:167 Uncaught ReferenceError: receiverId is not defined

First of all, try to open app.js and find the line 167. It can help to understand the issue better.

My guess is that you do not define `receiverId` before  window.Echo.private(`private-chat.${receiverId}`) call.

Let's take a look at your Bootstrap.js:

```javascript
window.Echo.private(`private-chat.${receiverId}`)
    .listen('PrivateMessageEvent', (e) => {
		console.log(e);
});
```

I personally do not use such placeholder as ${receiverId}. Instead, I use a simple JS concatenation to pass a parameter to the channel, for example:

```javascript
window.Echo.private(`private-chat.` + receiverId)
    .listen('PrivateMessageEvent', (e) => {
		console.log(e);
});
```
And I would make sure that `receiverId` is defined before `window.Echo.private(`private-chat.` + receiverId)`.

Hope it helps.

Konstantin

**Comment by kkomelin** [comment]
Also, I would not call window.Echo.private(`private-chat.` + receiverId) in Bootstrap.js. I would call it when and where I can get the receiverId parameter. 
The receiverId can be a part of your page url or can be represented by a hidden element in your template, for example: 
```html
<input type="hidden" id="receiver_id" value="{{$receiver_id}}" />
```
In case you use hidden element, just get its value and pass to the Echo.private() method.
```javascript
var receiverId = document.getElementById('receiver_id').value;
window.Echo.private(`private-chat.` + receiverId)
    .listen('PrivateMessageEvent', (e) => {
		console.log(e);
});
```
Something like that.

To sum up, I understand your pain. Some points of the documentation are not so obvious.
Frankly, I also spent a lot of time trying to make Echo + Echo Server work my first time. But don't try to solve the problem alone. Ask the community.

**Comment by xparthx** [comment]
Hey @kkomelin,

Thanks for your reply.

1) I have placed that socket.io code into resources\assets\js\bootstrap.js which came out of the box with the Laravel. 
I wrote it there because it was written in Laravel's official doc:
https://laravel.com/docs/master/broadcasting#receiving-broadcasts

2) public/js is just the compiled file of the resources/assets/js/app.js using _Laravel mix_
Line 168 contains that error. Please see the screenshot.

![js](https://cloud.githubusercontent.com/assets/4241851/24871860/bae13210-1e38-11e7-9da8-6a3bf7d84815.PNG)

3) I have used placeholder ${receiverId} just because it was written there on Laravel's official doc. I had already tried with simple JS concatenation but it still didn't work.

4) Also, I would not call window.Echo.private('private-chat.' + receiverId) in Bootstrap.js. I would call it when and where I can get the receiverId parameter.
-> That Js error occurs as soon as I visit the page. The event which broadcast, get only triggers upon form submission. So I think that is the main issue. I think I should call this code upon submission of that form instead of calling it directly in the bootstrap.js:

```
window.Echo.private(`private-chat.${receiverId}`)
    .listen('PrivateMessageEvent', (e) => {
		console.log(e);
});
```
Thanks

**Comment by kkomelin** [comment]
> So I think that is the main issue. I think I should call this code upon submission of that form instead of calling it directly in the bootstrap.js:

Exactly. That's what I meant. 
And the Laravel docs do not instruct you to put `window.Echo.private()` to bootstrap.js, as far as I know. Moreover, they don't instruct you where to put this code at all. That's the problem of the docs. What is obvious for the author can be not so obvious for users.
Anyway, try to put `window.Echo.private()` where the receiverId is known and defined.
For instance, I run this code on chat page load for one of my apps.

**Comment by xparthx** [comment]
Yes, that was the main issue. That solved half of the issue as it can get now `receiverId `. 

Now everything works great without any error. But I'm not getting any output on the console while listing that event:

```
$('#group-chat').on('submit', function(e) {
	var form = $(this);
	e.preventDefault();
	$.ajax({
		url: form.attr('href'),
		method: 'POST',
		data: form.serialize(),
		success: function(data) {
			form.find('#chat-message').val('');
			window.Echo.private(`chat.${data}`)
			.listen('PrivateMessageEvent', (e) => {
				console.log('test'); // Doesn't print anything
			    });
		}
	});
});
```
Console output from the echo server

```
Channel: private-chat.93
Event: App\Events\PrivateMessageEvent
CHANNEL private-chat.93
[11:40:45 PM] - JOKGoYrrjyTgIsLZAAAV authenticated for: private-chat.93
[11:40:45 PM] - JOKGoYrrjyTgIsLZAAAV joined channel: private-chat.93
```

Thanks

**Comment by kkomelin** [comment]
It's probably necessary to explain how the event broadcasting works:
1) User A sends a chat message through the form submit (or Ajax).
2) Your laravel app (route) receives the submit request and initiates event broadcasting broadcast(...)->toOthers().
3) The event goes to the Laravel Queue which uses Redis pub/sub (let's call it queue as well).
4) Since Laravel Echo Server listens to the certain kind of events in the Redis queue and receives the event of the User A.
5) After receiving the event, Laravel Echo Server broadcasts the event to other clients.
6) Finally, User B receives the User A's message.

So, you should run the event listener `window.Echo.private()` for User B only, and it should be somewhere on the page load.
And your form submit route callback should broadcast() the event.

From your code sample it seems like you start listening to the event only after sending the first message. I assume that if you don't send any messages, you won't receive messages from other users.

Just in case, I recommend you re-reading Laravel Broadcasting docs from start to the end to strengthen your knowledge.

**Comment by xparthx** [comment]
Yes, it makes sense.

But the issue is if I listen for that event on page load instead of the AJAX callback then again I'll get that `receiverId` undefined error.

I have read Laravel Broadcasting doc many times but it still doesn't tell how to tackle this issue. I don't blame doc for that. No doc can explain how to do some specific task - it has to generalise and so it is.

Would be good if you can just explain verbally that what I'm doing wrong here. 

Appreciate your help.

Thanks

**Comment by xparthx** [comment]
ok, I got that working i.e listens for an event on page load without getting that `receiverId` undefined error.

```
var receiverId = null;

window.Echo.private(`chat.${receiverId}`)
    .listen('PrivateMessageEvent', (e) => {
        console.log(e); // Not getting excute
});
```

Console output from the echo server

```
[11:45:56 AM] - vmqOY60cgP4ASis7AAAG authenticated for: private-chat.null
[11:45:56 AM] - vmqOY60cgP4ASis7AAAG joined channel: private-chat.null
Channel: private-chat.6
Event: App\Events\PrivateMessageEvent
CHANNEL private-chat.6

```

So now the only issue that on the JS, the event is not getting listen.

Thanks

**Comment by kkomelin** [comment]
The problem is in your architecture. Why do you need this `receiverId`? To check the access rights?
I would create a model, called Chat, and use `chatId` instead of `receiverId`, so that the app could always access `chatID` and received subscribed users by this `chatId`. Something like that.

> Would be good if you can just explain verbally that what I'm doing wrong here.

First of all, I have no clue what you're trying to achieve. Secondly, I can't see your app code. So, it's difficult for me to say exactly. Moreover, it's your job and not mine. Correct?

That's all I can help with. Good luck with the project!
Konstantin

**Comment by xparthx** [comment]
Hello,

Finally, I got everything working in a correct way. You were right about my architecture. It was wrong. Now I know when/how to fire event and when/how to listen for that event.

Thanks a lot for your support so far.

**Comment by kkomelin** [comment]
Hi @xparthx,

You're welcome.
I'm glad you did it yourself. Congrats!

Best,
Konstantin

**Comment by paparyku** [comment]
paparyku in the buildin

**Comment by jcrawford** [comment]
@xparthx could you explain what you were doing wrong with the architecture? I am facing this exact same problem after reading through the docs a few times the solution is not clear.  I am trying to figure out how to make echo listen to a post.${postId}.comment channel which is a private channel but I have absolutely no clue how to access the postId since the receiver of the message is not committing any action, it should just listen on the channel where the postId is dynamic.

**Comment by xparthx** [comment]
Sorry right now I'm bit busy, so can't help much.

But you can get idea from my repo:
https://github.com/xparthx/laravel-realtime-chat

Thanks

**Comment by mentalrob** [comment]
afşkmsafşkmsafaşsf

**Comment by tlaverdure** [comment]
Hey there, this issue was opened a while ago. I'm going to close this issue, if this issue still exists, please open a new issue or open a Pull Request.
